### ISSUE TITLE
`#[expect(unused)]` false-positive `unfulfilled_lint_expectations` warning on implemented but unused traits

### ISSUE THREAD

**Issue by yuriks** [rust-lang/rust#152370]
### Code

```Rust
#[expect(unused)]
trait UnusedTrait {}

struct UsedStruct(u32);

impl UnusedTrait for UsedStruct {}

fn main() {
    let x = UsedStruct(12);
    println!("Hello World! {}", x.0);
}
```

### Current output

```Shell
warning: this lint expectation is unfulfilled
 --> src/main.rs:1:10
  |
1 | #[expect(unused)]
  |          ^^^^^^
  |
  = note: `#[warn(unfulfilled_lint_expectations)]` on by default

warning: `playground` (bin "playground") generated 1 warning
```

### Desired output

```Shell
Compilation with no warnings
```

### Rationale and extra context

The code above throws an `unfulfilled_lint_expectations` lint even though, if it were removed, the compiler would emit a `dead_code` lint for the trait.
This only happens when the trait is implemented for some type but ultimately still ends up unused. If there are no implementations at all then `expect` suppresses the lint correctly without a warning.

### Other cases

```Rust
// If `#[expect(unused)]` is commented out:
warning: trait `UnusedTrait` is never used
 --> src/main.rs:2:7
  |
2 | trait UnusedTrait {}
  |       ^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` (part of `#[warn(unused)]`) on by default

warning: `playground` (bin "playground") generated 1 warning
```

### Rust Version

```Shell
rustc 1.94.0-beta.2 (23a44d3c7 2026-01-25)
binary: rustc
commit-hash: 23a44d3c70448c08dc6a2fc13c1afceab49f2bb9
commit-date: 2026-01-25
host: x86_64-pc-windows-msvc
release: 1.94.0-beta.2
LLVM version: 21.1.8
```

### Anything else?

Playground link: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=1d6bd45080b177104331d8968d8812b8

**Comment by mu001999** [comment]
In fact we also have:
```rust
#[expect(dead_code)]
fn foo() {
    if true {
        foo();
    }
}
```

This will lead to similar false-positive warnings.

**Comment by joshtriplett** [comment]
It looks like #154377 fixes many of the cases we were concerned about. Thanks, @mu001999!

We'd be happy to see that go forward.

We're un-nominating this issue for now. Feel free to re-nominate if there are more issues that that PR doesn't cover, and if there are proposed fixes or requested semantics for possible fixes. Otherwise, this could be closed.

If someone in the future wants to write up a more precise specification of the behavior of the lint, that'd be awesome and we'd appreciate it, but it's a lint, so we're allowed to be fuzzy and try to hill-climb towards the right behavior, and we can always change it in the future.
